AppDashboard/lib/datastore_viewer.py

```diff
@@ -0,0 +1,637 @@
+    for prop_name, _ in index.Properties():
+      if not prop_name in entity.keys() or (
+          prop_name in entity.unindexed_properties()):
+        return 0
+      prop_vals = entity[prop_name]
+      if isinstance(prop_vals, (list)):
+        composite_index_value_count *= len(prop_vals)
+
+    # If this is an ancestor index we're going to duplicate all these index
+    # writes for every key in the hierarchy.  So if the entity key has no
+    # parent we write the index values once, if the entity key has a depth of
+    # 2 then we write the index values twice, and so on.
+    ancestor_count = 1  # A key is its own ancestor.
+    if index.HasAncestor():
+      key = entity.key().parent()
+      while key is not None:
+        ancestor_count += 1
+        key = key.parent()
+    return composite_index_value_count * ancestor_count
+
+  def _construct_url(self, remove=None, add=None):
+    """Returns a URL referencing the current resource with the same params.
+
+    For example, if the request URL is
+    "http://foo/bar?animal=cat&color=redirect" then
+    _construct_url(['animal'], {'vehicle': 'car'}) will return
+    "http://foo/bar?vehicle=car&color=redirect"
+
+    Args:
+      remove: A sequence of query parameters to remove from the query string.
+      add: A mapping of query parameters to add to the query string.
+
+    Returns:
+      A new query string suitable for use in "GET" requests.
+    """
+    remove = remove or []
+    add = add or {}
+    params = dict(self.request.params)
+    for arg in remove:
+      if arg in params:
+        del params[arg]
+
+    params.update(add)
+    return str('%s?%s' % (self.request.path,
+                          urllib.urlencode(sorted(params.iteritems()))))
+
+  @classmethod
+  def _get_entity_template_data(cls, ds_access, request_uri, kind, namespace,
+                                order, start):
+    """ Fetches template variables for datastore viewer page for a given kind.
+
+    Args:
+      ds_access: A DatastoreDistributed client.
+      request_uri: A string specifying the current request location.
+      kind: A string specifying the entity kind.
+      namespace: A string specifying the datastore namespace.
+      order: A string containing the name of the property to sorted the results
+        by. A "-" prefix indicates descending order e.g. "-age".
+      start: The number of initial entities to skip in the result set.
+    Returns:
+      A tuple containing headers, entities, and total entity count.
+    """
+    entities, total_entities = _get_entities(
+      ds_access, kind, namespace, order, start, cls.NUM_ENTITIES_PER_PAGE)
+
+    property_name_to_value = _property_name_to_value(entities)
+
+    headers = [{'name': property_name}
+               for property_name in sorted(property_name_to_value)]
+
+    template_entities = []
+    for entity in entities:
+      attributes = []
+      for property_name in sorted(property_name_to_value):
```

Comment:
since this is the second sort should we store the sorted one in a variable or use the headers list and set a temp variable property_name to header['name'] ? Or since we never use the list without sorting it at the assignment it could be `property_name_to_value = sorted(_property_name_to_value(entities))` or the method could return it sorted

I just would rather we don't sort the same thing twice in this method.